Highlight Crates a Track is in

MK-42 opened the pull request:
I've added some basic handling for lp: 1380467 (https://bugs.launchpad.net/mixxx/+bug/1380467)

This will save the last selected track from trackSelected signal of the library and use a delegate for the LibraryTreeView to highlight the crates that track is in.

Todo:
Find a way to repaint the widget after style-change (any hints on that?) for now scrolling the treeView brings it to live.
Apply the same for playlists

Some questions:
1) how to trigger the repaint the best/preferred way?
2) is it ok what I'm doing there? Adding a getter for CrateDao Just trying to get this working, any hints are welcome.

btw: what is dao short for? Maybe st. like dataAccessObject? Just curious, maybe makes it easier to read for me.

Commit:
Highlight Crates a Track is in. Still some work to do

Commit:
Rename delegate according to his function

daschuer commented on src/library/cratefeature.cpp:
```diff
@@ -705,3 +707,15 @@ QString CrateFeature::getRootViewHtml() const {
 }
+
+TrackPointer CrateFeature::getSelectedTrack() {
+    return this->m_pSelectedTrack;
```
"this->" is redundant

daschuer commented on src/library/cratehighlightdelegate.cpp:
```diff
@@ -0,0 +1,28 @@
+#include "cratefeature.h"
+
+
+void CrateHighlightDelegate::initStyleOption(QStyleOptionViewItem *option,
```
Please always aligne * and & with the type QStyleOptionViewItem* option.

daschuer commented on src/library/cratehighlightdelegate.cpp:
```diff
@@ -0,0 +1,28 @@
+
+    TreeItem *item = static_cast<TreeItem*>(index.internalPointer());
+    LibraryFeature* pFeature = static_cast<LibraryFeature*>(item->getFeature());
+    if (pFeature){
```
use a space before {

daschuer commented on src/library/cratehighlightdelegate.cpp:
```diff
@@ -0,0 +1,28 @@
+    QStyledItemDelegate::initStyleOption(option, index);
+    QStyleOptionViewItemV4 *optionV4 = qstyleoption_cast<QStyleOptionViewItemV4*>(option);
+
+    TreeItem *item = static_cast<TreeItem*>(index.internalPointer());
```
Are you sure that this cast will always be valid and != null?

MK-42 replied:
I got this from TreeItemModel::data(), where it's done the same way. But i've added a condition for index.isValid() before. Thats the way we are doing it there too.

daschuer commented on src/library/cratehighlightdelegate.cpp:
```diff
@@ -0,0 +1,28 @@
+    if (pFeature){
+        TrackPointer pTrack = pFeature->getSelectedTrack();
+        if (pTrack){
+            CrateFeature* pCrateFeature = static_cast<CrateFeature*>(pFeature);
```
this will != null if the assumption above is valid. I think you want to check if pFeature is a crate first. 
So you might want to use dynamic_cast. 

But in general it is better to avoid all these upcasting. This can be done, by moving the required functions to the base class. Implement them in a generic way and virtual override them in CrateFeature.

daschuer commented on src/library/dao/cratedao.cpp:
```diff
@@ -422,3 +422,17 @@ void CrateDAO::removeTracksFromCrates(const QList<int>& ids) {
+        LOG_FAILED_QUERY(query);
+        return false;
+    }
+    return query.next();
```
This requires a comment.

daschuer commented:
Hi Markus, 

thank you for working on this. 

Sorry I have no answer for your question. 
Can one else jump  in here? 
If not I am afraid you need to search the web for a solution. 

So far the code seams to be on the right track. 
I have added only some code style nit picks.

MK-42 commented:
Thanks for your comments. Ive changed it and will commit shortly.

The repaint-issue is harder than I thought. I've tried with dataChanged signal of the TreeItemModel but without success. For now the web gives no other solution for me I fear. But will continue searching.

daschuer commented on src/widget/wlibrarysidebar.cpp:
```diff
@@ -27,6 +27,9 @@ WLibrarySidebar::WLibrarySidebar(QWidget* parent)
     header()->setHorizontalScrollMode(QAbstractItemView::ScrollPerPixel);
+
+    m_delegate = new CrateHighlightDelegate();
+    this->setItemDelegate(m_delegate);
```
this-> is redundant.

daschuer commented:
When is initStyleOption() called? The init prefix makes me think that it is only called one time. 
You may add a breakpoint or a qDebug() to check it. 

It think you have to invalidate the index inside CrateFeature::slotTrackSelected. 

Than you probably have to override the paint() function to paint yourself or override displayText and change the style there.  

http://doc.qt.io/qt-4.8/qstyleditemdelegate.html

You may also search Github for qstyleditemdelegate to see how others use it.

Commit:
Highlighting crates the selected track is in works now

MK-42 commented:
I've managed to fix it. kind of hackish, but works. I'm telling the model that I will insert rows what gives a repaint. Nothing else worked for me.

If you're ok with this, I will also add handling for playlists before we merge this.

daschuer commented on src/library/cratehighlightdelegate.cpp:
```diff
@@ -0,0 +1,27 @@
+#include "cratefeature.h"
+
+
+void CrateHighlightDelegate::initStyleOption(QStyleOptionViewItem* option,
```
Can you add here a comment, when it is called?

Commit:
Some minor code-style fixes

Commit:
Added comment to CrateHighlightDelegate::initStyleOption

daschuer commented on src/library/treeitemmodel.cpp:
```diff
@@ -193,3 +193,9 @@ TreeItem* TreeItemModel::getItem(const QModelIndex &index) const {
 }
+
+void TreeItemModel::triggerRepaint() {
+    //todo: find a nicer way to do this. Works as it should but kind of hackish
```
did you try 
emit dataChanged() 
http://doc.qt.io/qt-5/qabstractitemmodel.html#dataChanged
"When reimplementing the setData() function, this signal must be emitted explicitly."

QModelIndex().parent() looks a bit scary can we use at least the real indexes here? 


This function repaints the whole tree, for some extra points, it would be nice to invalidate only the effected items.

MK-42 replied:
Yes, i tried emit dataChanged() with various parameters. I couldn't get it to work with that. From that point on, I was unable to repaint only some items so I stick with repainting the whole tree.

rryan replied:
This is similar to how we do it in BaseSqlTableModel. Did you try this way?
```c++
QModelIndex left = index(0, 0);
QModelIndex right = index(numRows, numColumns);
emit(dataChanged(left, right));
```

MK-42 replied:
Yes I did (just now once again to be sure). It has no effect. The Widget will not repaint.

I have a solution now without the sideeffect of moving selection by doing
`beginInsertRows(QModelIndex(), rowCount(), rowCount());`
`endInsertRows()`

but that also isn't the nice way...

rryan replied:
I figured this one out -- SidebarModel wasn't re-broadcasting the dataChanged signal.

daschuer commented:
> If you're ok with this, I will also add handling for playlists before we merge this.

Yes, go ahead.

Commit:
Added highlighting of playlists

MK-42 commented:
well... playlists are buggy. For some reason, I do not understand, the selection in the treeview moves one item down if a playlist is selected. Will investigate...

MK-42 commented:
Oh. actually that happens for crates also. So we need another way to trigger the repaint than that hack, that seems to have side-effects.

Commit:
Another aproach for Repainting preventing the selection to move

rryan commented:
Thanks for working on this Markus. 

I'm worried that doing blocking database queries in response to UI movements will make it jerky for rapid scrolling through the library. Also, this implementation does one database query per crate/playlist when you could do the same with 2 queries (one to get all the crates a track is in and one to get all the playlists a track is in). 

A couple ways this could be resolved:
* A track could have a list of playlist and crates it is a member of that is maintained by the TrackInfoObject. (Hard because PlaylistDao / CrateDao do not have access to the TIO currently). 
* PlaylistDao/CrateDao could keep an in-memory map of track id -> [playlist/crate id], that represents the state of the database for fast lookups.

MK-42 commented:
would you prefer  to go with the in memory map or doing it with two querys, by holding the crates/playlists a track is in in memory?
The second one would mean 2 Queries each selection move, the first would need one initializing and none after that by handling all the signals of the DAOs i think.

When we decide to do it in memory, which dataStructure would be preferred? using
`List<Pair<int, List<int> > >` seems ugly :D

Still no nice solution for repainting, but one that works without sideeffects so far.

rryan commented:
> would you prefer to go with the in memory map or doing it with two querys, by holding the crates/playlists a track is in in memory?
> The second one would mean 2 Queries each selection move, the first would need one initializing and none after that by handling all the signals of the DAOs i think.

I think it'd be best to go with the in-memory map idea. That way we only pay a small insert/update cost on every Playlist/Crate modification operation in the DAOs. Compared to the DB queries associated with each of those operations the extra cost is tiny :).

Whereas doing queries in response to GUI operations has caused a lot of jerkiness in our UIs in the past.

> When we decide to do it in memory, which dataStructure would be preferred? using
> List<Pair<int, List<int> > > seems ugly :D

A QMultiHash<int, int> is probably best here. That provides constant-time lookup.

rryan commented:
Hm, I'm getting segfaults on start:

```
Warning [Main]: ControlDoublePrivate::getControl returning NULL for ( "[EffectRack1]" , "show" )
Warning [Main]: Requested control does not exist: "[EffectRack1],show" Creating it.
Warning [Main]: Object::connect: No such signal MixxxLibraryFeature::trackSelected(TrackPointer) in src/library/mixxxlibraryfeature.cpp:131
Warning [Main]: Object::connect:  (sender name:   'DlgHidden')
Warning [Main]: Object::connect: No such signal MixxxLibraryFeature::trackSelected(TrackPointer) in src/library/mixxxlibraryfeature.cpp:137
Warning [Main]: Object::connect:  (sender name:   'DlgMissing')
Warning [Main]: Object::connect: No such signal AutoDJFeature::trackSelected(TrackPointer) in src/library/autodj/autodjfeature.cpp:123
Warning [Main]: Object::connect:  (sender name:   'DlgAutoDJ')
Debug [Main]: Recordings folder set to "/Users/rjryan/Music/Mixxx/Recordings"
Debug [BrowseThread]: Append last  0
Debug [Main]: BaseTrackCache(0x11d621a50) updateIndexWithQuery took 4 ms
Debug [Main]: AnalysisLibraryTableModel(0x15a3b4db0) select() took 12 ms 0
Debug [Main]: AnalysisLibraryTableModel(0x15a3b4db0) select() took 2 ms 0
Warning [Main]: Object::connect: No such signal AnalysisFeature::trackSelected(TrackPointer) in src/library/analysisfeature.cpp:76
Warning [Main]: Object::connect:  (sender name:   'DlgAnalysis')
Debug [Main]: DlgAnalysis(0x15a33e320, name = "DlgAnalysis") analysisActive false
Warning [Main]: ControlDoublePrivate::getControl returning NULL for ( "[VinylControl]" , "show_vinylcontrol" )
Debug [Main]: Setting visibility for "[PreviewDeck]" "show_previewdeck" true
Warning [Main]: ControlDoublePrivate::getControl returning NULL for ( "[VinylControl]" , "show_vinylcontrol" )
Process 16497 stopped
* thread #1: tid = 0x4847f2, 0x00000001003dcd41 mixxx`CrateHighlightDelegate::initStyleOption(this=<unavailable>, option=<unavailable>, index=<unavailable>) const + 113 at cratehighlightdelegate.cpp:25, queue = 'com.apple.main-thread', stop reason = EXC_BAD_ACCESS (code=EXC_I386_GPFLT)
    frame #0: 0x00000001003dcd41 mixxx`CrateHighlightDelegate::initStyleOption(this=<unavailable>, option=<unavailable>, index=<unavailable>) const + 113 at cratehighlightdelegate.cpp:25
   22  	    TreeItem *item = static_cast<TreeItem*>(index.internalPointer());
   23  	    LibraryFeature* pFeature = item->getFeature();
   24  	    if (pFeature) {
-> 25  	        TrackPointer pTrack = pFeature->getSelectedTrack();
   26  	        if (pTrack) {
   27  	            if (pFeature->isTrackInChildModel(pTrack->getId(), item->dataPath())){
   28  	                optionV4->font.setBold(true);
(lldb) bt all
* thread #1: tid = 0x4847f2, 0x00000001003dcd41 mixxx`CrateHighlightDelegate::initStyleOption(this=<unavailable>, option=<unavailable>, index=<unavailable>) const + 113 at cratehighlightdelegate.cpp:25, queue = 'com.apple.main-thread', stop reason = EXC_BAD_ACCESS (code=EXC_I386_GPFLT)
  * frame #0: 0x00000001003dcd41 mixxx`CrateHighlightDelegate::initStyleOption(this=<unavailable>, option=<unavailable>, index=<unavailable>) const + 113 at cratehighlightdelegate.cpp:25
    frame #1: 0x0000000101826d7e QtGui`QStyledItemDelegate::sizeHint(QStyleOptionViewItem const&, QModelIndex const&) const + 158
    frame #2: 0x00000001017b2c39 QtGui`QTreeView::indexRowSizeHint(QModelIndex const&) const + 1193
    frame #3: 0x00000001017b2d63 QtGui`QTreeViewPrivate::itemHeight(int) const + 115
    frame #4: 0x00000001017b2f0f QtGui`QTreeViewPrivate::updateScrollBars() + 207
    frame #5: 0x00000001017c013d QtGui`QTreeView::updateGeometries() + 477
    frame #6: 0x000000010176ccde QtGui`QAbstractItemView::doItemsLayout() + 46
    frame #7: 0x00000001017ba72e QtGui`QTreeView::doItemsLayout() + 766
    frame #8: 0x00000001017b4f91 QtGui`QTreeView::scrollTo(QModelIndex const&, QAbstractItemView::ScrollHint) + 113
    frame #9: 0x0000000100643870 mixxx`WLibrarySidebar::selectIndex(this=0x000000015a2d7840, index=0x00007fff5fbfeb70) + 320 at wlibrarysidebar.cpp:209
    frame #10: 0x0000000100fdb4b2 QtCore`QMetaObject::activate(QObject*, QMetaObject const*, int, void**) + 2498
    frame #11: 0x0000000100486623 mixxx`SidebarModel::selectIndex(this=<unavailable>, _t1=<unavailable>) + 51 at moc_sidebarmodel.cc:142
    frame #12: 0x00000001004c4dab mixxx`SidebarModel::activateDefaultSelection(this=0x000000011beeab70) + 107 at sidebarmodel.cpp:60
    frame #13: 0x0000000100fdb4b2 QtCore`QMetaObject::activate(QObject*, QMetaObject const*, int, void**) + 2498
    frame #14: 0x00000001004f1d7a mixxx`MixxxMainWindow::MixxxMainWindow(this=0x00000001032828c0, pApp=<unavailable>, args=0x0000000100a1e060) + 7914 at mixxx.cpp:402
    frame #15: 0x00000001004de817 mixxx`main(argc=<unavailable>, argv=<unavailable>) + 1351 at main.cpp:304
    frame #16: 0x00007fff896705c9 libdyld.dylib`start + 1
    frame #17: 0x00007fff896705c9 libdyld.dylib`start + 1
```

MK-42 commented:
thats's weird... I'm on linux, everything is working here...

rryan commented on src/library/cratehighlightdelegate.cpp:
```diff
@@ -0,0 +1,32 @@
+    QStyledItemDelegate::initStyleOption(option, index);
+    QStyleOptionViewItemV4 *optionV4 = qstyleoption_cast<QStyleOptionViewItemV4*>(option);
+
+    TreeItem *item = static_cast<TreeItem*>(index.internalPointer());
```
The segfault is because the internal pointers in the sidebar can be ```SidebarModel*``` for items that are children of the root and ```TreeItem*``` for all others.

MK-42 replied:
I'm not that experienced with that casting.
So I should do a dynamic_cast here and check if the result is ok?

rryan replied:
Sent you a PR :).

rryan committed:
Ignore QModelIndex's that don't have TreeItem internal pointers in CrateHighlightDelegate.

rryan committed:
Fix SidebarModel handling of child models' dataChanged signals.

MK-42 committed:
Merge pull request #1 from rryan/sidebar-highlight

Sidebar highlight fixes.

MK-42 commented:
So I'll go on with the QMultiHash in the daos. Thanks :+1:

MK-42 commented:
Should I do the online adding and removing in slots connected to the corresponding signals of the dao or directly in the methods?

rryan commented:
> Should I do the online adding and removing in slots connected to the corresponding signals of the dao or directly in the methods?

I wouldn't do it in a slot since then there's no guarantee your slot runs before some other slot listening to signals from the PlaylistDao or CrateDao.

Commit:
Holding Track-Crate relation in memory for faster highlighting

Commit:
Holding Track-Playlist relation in memory for faster highlighting

MK-42 commented:
Everything in memory now ;-)
Fetching once at startup. But I can't prevent this one ;-)

rryan commented on src/library/baseplaylistfeature.cpp:
```diff
@@ -82,6 +83,10 @@ BasePlaylistFeature::BasePlaylistFeature(QObject* parent,
     connect(&m_playlistDao, SIGNAL(lockChanged(int)),
             this, SLOT(slotPlaylistTableChanged(int)));
+
+    Library* pLibrary = static_cast<Library*>(parent);
```
I think this is redundant with library.cpp:229 -- all LibraryFeatures should already get this signal. Did you find that it wasn't being delivered?

rryan replied:
Whoops, I got that backwards.

rryan commented on src/widget/wlibrarysidebar.cpp:
```diff
@@ -27,6 +27,9 @@ WLibrarySidebar::WLibrarySidebar(QWidget* parent)
+    setItemDelegate(m_delegate);
 }
 
 WLibrarySidebar::~WLibrarySidebar() {
```
should ```delete m_delegate;``` here since setItemDelegate doesn't take ownership

Commit:
Clean up used memory

rryan commented:
It looks like this doesn't quite work in the special views -- hidden, autodj, analyze, history. I think fixing these signal connections will fix it:

Warning [Main]: Object::connect: No such signal MixxxLibraryFeature::trackSelected(TrackPointer) in src/library/mixxxlibraryfeature.cpp:131
Warning [Main]: Object::connect:  (sender name:   'DlgHidden')
Warning [Main]: Object::connect: No such signal MixxxLibraryFeature::trackSelected(TrackPointer) in src/library/mixxxlibraryfeature.cpp:137
Warning [Main]: Object::connect:  (sender name:   'DlgMissing')
Warning [Main]: Object::connect: No such signal AutoDJFeature::trackSelected(TrackPointer) in src/library/autodj/autodjfeature.cpp:123
Warning [Main]: Object::connect:  (sender name:   'DlgAutoDJ')
Warning [Main]: Object::connect: No such signal AnalysisFeature::trackSelected(TrackPointer) in src/library/analysisfeature.cpp:76
Warning [Main]: Object::connect:  (sender name:   'DlgAnalysis')

rryan commented:
I wonder how multiple selected tracks should work? That's a bigger job than this PR since the trackSelected signal you're using inherently only deals with one track. That's because cover art only wants the most recently selected track which was what htat signal was initially created for. Oh well :).

MK-42 commented:
> I wonder how multiple selected tracks should work? 

Would it be ok to just include that version for now? I doubt I can do the other one for now.

> It looks like this doesn't quite work in the special views

Maybe call me dumb, but I dont't understand those. I've never touched these connects and didn't introduced new signals anywhere, or did I?

rryan commented:
> I wonder how multiple selected tracks should work?
> Would it be ok to just include that version for now? I doubt I can do the other one for now.

Yea, that's what I was saying :). No need to handle in this PR since the signals you're using are kind of inherently 1-track oriented.

> It looks like this doesn't quite work in the special views
> Maybe call me dumb, but I dont't understand those. I've never touched these connects and didn't introduced new signals anywhere, or did I?

Yea, I think those were broken in master as well.

rryan commented:
Yea, adding a  ```void trackSelected(TrackPointer pTrack);``` signal to each of AutoDJFeature/AnalysisFeature/MixxxLibraryFeature/etc. should fix it.

rryan commented:
Hm, another corner case -- selecting a different view doesn't clear the bolding. Since cover art clears when you switch views we should probably clear the bolding as well for consistency.

Commit:
fix missing signals

Commit:
Clear selected Playlists/Crates on viewChange

MK-42 commented:
> Hm, another corner case

Is now handled

daschuer commented on src/library/dao/cratedao.cpp:
```diff
@@ -18,6 +18,32 @@ CrateDAO::~CrateDAO() {
+
+    //get the count to allocate HashMap
+    int tracksInCratesCount = 0;
+    QSqlQuery query(m_database);
```
does query.size() return a valid value with sqlite? Than we can get around the second query.

If not, you need not Query if  tracksInCratesCount == 0

MK-42 replied:
according to http://stackoverflow.com/questions/5373184/qtsql-sqlite-and-support-for-size-function it doesn't.

daschuer commented on src/library/dao/cratedao.cpp:
```diff
@@ -18,6 +18,32 @@ CrateDAO::~CrateDAO() {
+    while (query.next()) {
+        m_cratesTrackIsIn.insert(query.value(crateIdColumn).toInt(),
+                                 query.value(trackIdColumn).toInt());
+    }
```
This basically copies the whole crates table to memory. 
Up to now we have to care about this table and the original library table. 
This is extra work for all crate actions and a source of errors because of two data storages. 

Do we have test data, how this compares to the original solution?
I think we have a penalty on mass actions.
Is it worth the work? Remember that the selected Track slot is already decoupled by a timer.   

Is there an alternative way to speed up the sql lookups e.g. a temp table?

rryan replied:
> This basically copies the whole crates table to memory. 
> Up to now we have to care about this table and the original library table. 
> This is extra work for all crate actions and a source of errors because of two data storages.

That's right -- but it's not that much memory (roughly 16-bytes per tuple so we're talking about like 16MB if every track was in a playlist in a 1-million track library). The main issue is keeping them in sync and since the DAO is already the gate-keeper it will do that fine.

I think having an in-memory index for fast lookups will help for bugs like these too:
https://bugs.launchpad.net/mixxx/+bug/1402133
https://bugs.launchpad.net/mixxx/+bug/1021422

For a GMail-style label delegate in the main library view we will need a fast lookup in memory because I can only think of 2 other ways:
* Doing an additional query per visible track in the UI.
* Incorporating a join of all crates/playlists into the root library view. 

Going with option 1 would mean terrible things for UI performance. Going with option 2 would mean adding a track to a crate would invalidate / recompute the library SQLite view which I don't think would be speedy.

rryan replied:
> Do we have test data, how this compares to the original solution?
> I think we have a penalty on mass actions.
> Is it worth the work? Remember that the selected Track slot is already decoupled by a timer.

It's currently hard to capture UI jankiness in a test. But it blocks the UI thread -- that's for sure. Until we can merge Nazar's work and keep the UI thread unblocked when doing DB operations then I think we should go this route.

MK-42 replied:
Well, which way to go?
For this one we could also query the crates/Playlists the selected track is in in slotTrackSelected and look up there. But that won't solve the other problem for the two mentioned bugs above.

daschuer replied:
This is finally a decision of taste. If you have a working solution that has a reasonable CPU load it is a good one. 
It is no issue to change the way of implementation later if we actually found performance issues. 

This and to quoted bugs can be solved with an in memory table and with an sql qerry. 

Outlook for the future: 
Nazars branch provides a non blocking access to sql database and will be part of Mixxx 1.13. The result is prepared in a separate thread and passed to the GUI thread in a new slot. 
Once this is integrated we have the option to block the GUI until the result is calculated from an in memory table or delay the GUI update in a non blocking way by until the SQL query is finished. 
Such a decoration feature like this is IMHO a candidate for the later option. 

Without the non blocking SQL access, an internal table access will block the GUI shorter, than an sql query, but we need more CPU to keep the table up to date and there is an amount of additional code maintenance and complexity. For now it looks like the internal table wins. 
Do you have a working solution for both? Take the on that suits more to you.   
I have no Idea which performance tweaks are worth the works. 

I think I will test it on my atom net-book, once we consider this as merge-able.     
  
By the way: Nazars branch needs a new maintainer. The latest version is already working, but it falls behind the master branch. The next steps to do is to merge it with master by crawling up the master branch one by one database related commit. You will find it here: https://github.com/mixxxdj/mixxx/pull/73

MK-42 replied:
> This is finally a decision of taste. If you have a working solution that has a reasonable CPU load it is a good one. 

At the moment this branch contains a working solution. I've worked with it a while without performance problems. I'm on an i5 with 16Gig RAM so your atom would be the better test-case. 

> Take the on that suits more to you.
As the in-memory table is ready now and tested for me, I would stick with that one for now. Once we have Nazars work in it would be worth refactoring that though - having multiple datasets that have to be in sync is clearly a point of failure.

So what about a test on your atom platform and when scolling by holding down arrowkeys or sth. like this isn't slowed down merge this as is and else test the other solution to see whats better?

I've just looked in nazars branch... will take a while to get into this...

daschuer commented on src/library/cratehighlightdelegate.cpp:
```diff
@@ -0,0 +1,45 @@
+        return;
+    }
+
+    if (pFeature->isTrackInChildModel(pTrack->getId(), item->dataPath())){
```
This leads finally to a conains() call for every crate which end up in while loop that compare key and item.
We should think about an sql Query a similar operation on an optimized hash table that is done one time in the selected track slot, preparing a list of crates that are highlighted.

daschuer replied:
Of cause the "is it worth" question should be solved first.

rryan replied:
It's a little tricky due to the structure of the delegate -- but yes that would be better. 

This could be done without a delegate by changing TreeItemModel::data() to return a bolded font for the Qt::FontRole if the crate is selected. That way you could get the list of crates/playlists that should be bold there and then access the list in TreeItemModel::data directly.

MK-42 replied:
> This could be done without a delegate by changing TreeItemModel::data() to return a bolded font for the Qt::FontRole if the crate is selected.

I had no luck with those roles, nothing changed if I used them. So I chose the delegate.

rryan replied:
I think this is also a case where SidebarModel -> TreeItemModel is tripping things up. I'll send you a PR

rryan committed:
Remove CrateHighlightDelegate.

rryan committed:
Add bold attribute to TreeItem.

rryan committed:
Restore LibraryFeature::trackSelected signal.

I think this was why we had to add signals to each individual feature.

rryan committed:
Restore LibraryFeature::trackSelected signal.

I think this was why we had to add signals to each individual feature.

rryan committed:
Respond to Qt::FontRole in SidebarModel.

rryan committed:
Populate playlist and crate TreeItems with bold attribute when track selection changes.

rryan committed:
Populate TreeItem state when re-building the child model.

This supports bolding/unbolding the playlist/crate when a track is added/removed.

MK-42 committed:
Merge pull request #2 from rryan/sidebar-highlight

Some fixes.

daschuer commented:
I have just tested the branch on my netbook and it works like a charm without any notable delays. 

On such small screen the tree is most of the time collapsed. So it would be nice to have the playlist and crate root item bold as well. Can this be part of this PR?

daschuer commented:
@rryan: 
I would like to do a 1.12.1 bugfix release after releasing 1.12.0 including the seek and floating point fixes that are currently part of master. How is the best way to proceed here? 
a) merge it to master as part of 1.12.1
b) merge it to master after we cut a 1.12.1 branch
c) merge master to 1.12.0 and ditch the 1.12.1 idea (due to pending windows issues), merge this to master afterwards. 

IMHO the 1.12 release is blocked by the windows crashes. 
How do you think about installing the pdb debugging file on windows? 
Alternative it would be nice if you could provide the pdb file as separate download.
How big is the file? Microsoft provides some options to strip it to a reasonable size.

daschuer commented:
By the way: LGTM 
Thank you @MK-42 and @rryan for this nice feature.

MK-42 commented:
>I have just tested the branch on my netbook and it works like a charm without any notable delays.

Did you test the branch with the last changes of RRyan? Those are slower for me in setlog-feature

daschuer commented:
How dis you test exactly? 
I have just tested the Crate add performance and the library scroll performance
Both works for me in the current state. I have not compared different solutions in this branch.

MK-42 commented:
select a track and hold down arrow keys to move the selection. Before merging rryans changes (dc1f573d27e3b972167a17804b7ce10ab9866076) this wasn't slowed down for me in the "SetLogFeature (History)", now the selection move very slow about one move per second. So I acctually would prefer to revert his changes.

Sorry that I didn't checked that before merging his changes, those looked better than my approach. But for me his changes are verry slow.

rryan commented:
Hmm, the change should be the same in terms of number of calls to the dao
(linear). It just removed the delegate. I'll take a look.

On Sun, May 31, 2015, 10:55 AM MK-42 <notifications@github.com> wrote:

> select a track and hold down arrow keys to move the selection. Before
> merging rryans changes (dc1f573
> <https://github.com/mixxxdj/mixxx/commit/dc1f573d27e3b972167a17804b7ce10ab9866076>)
> this wasn't slowed down for me in the "SetLogFeature (History)", now the
> selection move very slow about one move per second. So I acctually would
> prefer to revert his changes.
>
> Sorry that I didn't checked that before merging his changes, those looked
> better than my approach. But for me his changes are verry slow.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/mixxxdj/mixxx/pull/598#issuecomment-107200687>.
>

MK-42 commented:
> It just removed the delegate. I'll take a look.
Maybe it is because I am only quering the delegate (and for that the dao) on repainting the visible items and your approach for every element in the treeview? But I am not sure about this.

daschuer commented:
I cannot confirm the scrolling issue in a setlog playlsist, on my netbook.

Maybe this is due to the fact that the selected track slot is always delayed anyway on this slow device.

MK-42 commented:
would it be possible for you to test 
dc1f573d27e3b972167a17804b7ce10ab9866076
as well to see if it is faster on your device too? On mine it is.

MK-42 commented:
It gets slow if i expand the History-Item to show my about 300 History-Entries.

daschuer commented:
cannot see difference with my 20 entries for both commits.

MK-42 commented:
@rryan would it be OK to go back to the Delegate solution then, because it was faster for me?

rryan commented:
I'd like to avoid delegates as they prevent us from taking faster paths in the Qt library code (generally we've had tons of trouble with the 3 delegates in our library table causing performance issues).

rryan commented:
> @rryan would it be OK to go back to the Delegate solution then, because it was faster for me?

I think reversing the playlist hash key/value pairing to be (track, playlist) instead of (playlist, track) will help. As it works now we do a contains check for every playlist when you select a new track. Because ```QHash<int, int>::contains(int, int)``` does a hash lookup and then linear search through all the values under the key, it first does a constant time hash fetch on the playlist ID and then iterates through every element that matched (i.e. all the tracks in the playlist). So every contains check actually iterates through the whole list of tracks in the playlist. 

If the key/values were switched then it would be more efficient since we would typically be iterating through only the playlists IDs the track is in. 

Another downside is that we do N hash lookups to iterate through the playlists the track is in. Since we have a multi-hash we could actually just search through the iterator. Instead of offering a contains() check from PlaylistDAO we should probably instead offer a lookup method that returns an iterator or maybe just populates a list or a set.

rryan commented:
I'm happy to make the above changes and then maybe you could test if they fix your slowness @MK-42?

MK-42 commented:
> then maybe you could test if they fix your slowness @MK-42?

I can test that, no problem :+1:

rryan commented:
> then maybe you could test if they fix your slowness @MK-42?
> I can test that, no problem

Sent you a PR

MK-42 commented:
Sorry, I'm busy. Will test this tonight

Am 04.06.2015 um 17:04 schrieb RJ Ryan:
>     then maybe you could test if they fix your slowness @MK-42
>     <https://github.com/MK-42>?
>     I can test that, no problem
> 
> Sent you a PR
> 
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/mixxxdj/mixxx/pull/598#issuecomment-108927402>.
>

MK-42 commented:
As commented in that pr, thats not faster and i have other playlists/tracks highlighted then those that contain the tracks. So highlighting is wrong.
I will now test again performance with my delegate solution and if that works will revert to that position. I think we should not merge the state as is, because its slow for me.

MK-42 committed:
Revert "Some fixes."

MK-42 committed:
Merge pull request #4 from MK-42/revert-2-sidebar-highlight

Revert "Some fixes." these were slow

MK-42 commented:
Works for me now as expected - hovers the right things and is fast. ok to merge for me

rryan commented:
I'd prefer to iterate on the previous commit since avoids adding extra code and avoids a delegate. (the 3 delegates in our library table are the single biggest source of scrolling smoothness issues in the library).

The wrong data was a dumb mistake -- just pushed a commit to my branch:
https://github.com/rryan/mixxx/commit/41c5fa5965a2828ebcf66e6ede51fc87c8ae0689

MK-42 commented:
I've just seen that i missed an important change in your changed and cherry-picked that. But will test your branch now once again.

MK-42 commented:
Well. Now it does highlighting right. But as soon as I expand the 'History' LibraryFeature, it takes about 2 seconds to move the selection if I press a arrow-key on my keyboard. The same with selecting one entry with the mouse. It takes about that time to get the selection change. Like this this is nearly unusable. One could deactivate bolding in the history-tree, but I would actually like the feature to see "did i played that the last time?".

Could it be, that the delegate only fetches the 'isBold'-data for the visible items and the current solution for all of them?

rryan commented:
> Well. Now it does highlighting right. But as soon as I expand the 'History' LibraryFeature, it takes about 2 seconds to move the selection if I press a arrow-key on my keyboard. The same with selecting one entry with the mouse. It takes about that time to get the selection change. Like this this is nearly unusable. One could deactivate bolding in the history-tree, but I would actually like the feature to see "did i played that the last time?".
> 
> Could it be, that the delegate only fetches the 'isBold'-data for the visible items and the current solution for all of them?

Yea, this change populates the bold data once when the track is selected for all playlists.

To verify that this method is taking a long time, could you stick a PerformanceTimer in it? For example, stick this at the top of BasePlaylistFeature::slotTrackSelected:

```c++
ScopedTimer t("BasePlaylistFeature::slotTrackSelected");
```

Then run in --developer mode, select some tracks with the History section open, and then quit. It will print out a stats page to the console including statistics about the run time of that method.

How many history playlists do you have?

MK-42 commented:
> How many history playlists do you have?

about 300 i think. It's not that comfortable to delete them (there is no mass-delete or is there?) so they just build up...

```
Debug [Main]: Stat("BasePlaylistFeature::slotTrackSelected","count=18,sum=1.1583e+10ns,average=6.43498e+08ns,min=40365ns,max=1.49523e+09ns,variance=5.48256e+17ns^2,stddev=7.40443e+08ns")
```

6.4e+08ns is 0.6 seconds. It feels longer than that for the selection to change but I think that there is the problem then?

MK-42 commented:
The DAO seems fast:
```
Debug [Main]: Stat("PlaylistDAO::getPlaylistsTrackIsIn","count=16,sum=39816ns,average=2488.5ns,min=171ns,max=5270ns,variance=2.41036e+06ns^2,stddev=1552.53ns")
```

so I think it is the for-loop after that setting the bold-info or am I missing something?
Ive created a new block around the loop with {} and put a timer in there:
```
Debug [Main]: Stat("BasePlaylistFeature::slotTrackSelected","count=18,sum=1.13625e+10ns,average=6.31248e+08ns,min=42621ns,max=1.44179e+09ns,variance=5.27324e+17ns^2,stddev=7.26171e+08ns")
Debug [Main]: Stat("BasePlaylistFeature::slotTrackSelected_InLoop","count=18,sum=1.13623e+10ns,average=6.31242e+08ns,min=35990ns,max=1.44178e+09ns,variance=5.27325e+17ns^2,stddev=7.26171e+08ns")
```
There we are - its nearly the whole time in that loop.

MK-42 commented:
I think the problem is the setData in TreeItemModel which is called hundreds of times and everytime triggers a repaint via emit dataChanged.
I looked for something like `beginUpdate` and `endUpdate` but with no success

rryan commented:
@MK-42 ahh thanks for digging into that! Makes sense. To sidestep the updates we can reach into the TreeItem model itself:

```c++
diff --git a/src/library/baseplaylistfeature.cpp b/src/library/baseplaylistfeature.cpp
index 553e381..550f1d3 100644
--- a/src/library/baseplaylistfeature.cpp
+++ b/src/library/baseplaylistfeature.cpp
@@ -600,16 +600,24 @@ void BasePlaylistFeature::slotTrackSelected(TrackPointer pTrack) {
     int trackId = pTrack.isNull() ? -1 : pTrack->getId();
     m_playlistDao.getPlaylistsTrackIsIn(trackId, &m_playlistsSelectedTrackIsIn);

+    TreeItem* rootItem = m_childModel.getItem(QModelIndex());
+    if (rootItem == NULL) {
+        return;
+    }
+
     // Set all playlists the track is in bold (or if there is no track selected,
     // clear all the bolding).
     int row = 0;
     for (QList<QPair<int, QString> >::const_iterator it = m_playlistList.begin();
          it != m_playlistList.end(); ++it, ++row) {
         int playlistId = it->first;
-        QModelIndex index = m_childModel.index(row, 0);

+        TreeItem* playlist = rootItem->child(row);
+        if (playlist == NULL) {
+            continue;
+        }
         bool shouldBold = m_playlistsSelectedTrackIsIn.contains(playlistId);
-        m_childModel.setData(index, shouldBold, TreeItemModel::kBoldRole);
+        playlist->setBold(shouldBold);
     }
 }
```

rryan commented:
But then it needs a full dataChanged refresh using a method like the one you wrote.

rryan commented:
> But then it needs a full dataChanged refresh using a method like the one you wrote.

TreeItemModel::triggerRepaint

MK-42 commented:
Thats it. Like that it works. Thanks :)

But now ive screwed up that thing so much. Would it be ok to do a PR against your branch and you merge yours into master?

rryan commented:
nice! sure I'll commit that patch and make a PR

MK-42 commented:
closing this as @rryan opend another one for this feature.
